src/sys/windows/named_pipe.rs

```diff
@@ -645,7 +649,7 @@ fn write_done(status: &OVERLAPPED_ENTRY, events: Option<&mut Vec<Event>>) {
     // Acquire the `Arc<Inner>`. Note that we should be guaranteed that
     // the refcount is available to us due to the `mem::forget` in
     // `schedule_write` above.
-    let me = unsafe { overlapped2arc!(status.overlapped(), Inner, write) };
+    let me = unsafe { Arc::from_raw((status.overlapped() as usize - WRITE_OFFSET) as *mut Inner) };
```

Comment:
I would prefer going through `*mut u8` here. Integer to pointer casts are sketchy soundness-wise.
```suggestion
    let me = unsafe { Arc::from_raw(status.overlapped().cast::<u8>().sub(WRITE_OFFSET).cast::<Inner>()) };
```
or if the constants are 0, 1, 2, you can even do this.
```suggestion
    let me = unsafe { Arc::from_raw(status.overlapped().sub(WRITE_OFFSET).cast::<Inner>()) };
```

Reply:
Any reason not to use `wrapping_sub` which is less subtle to call?

Reply:
I agree the code is somewhat sketchy, but that really the nature of this code. Going via `sub` is a mistake as that uses `T` * count. Even if you cast to a `u8`, like you suggested, it's too easy to mess up imo.

Reply:
You could define helper functions for wrapping the subtraction and test them in the test :woman_shrugging:

Reply:
@Darksonn I've replaced the constants with methods and updated the test above.